src/lvm.c

```diff
@@ -0,0 +1,290 @@
+        continue;
+      }
+
+      char *lv_attr = get_json_string(lv, "lv_attr");
+      if (!lv_attr)
+        continue; // get_json_string() would have emitted an error message
+
+      // Skip virtual/thin LVs that don't use actual space in the VG
+      if ('v' == lv_attr[0] || 'V' == lv_attr[0])
+        continue;
+
+      // Report usage within thin pools
+      if ('t' == lv_attr[0]) {
+        char *lv_name = get_json_string(lv, "lv_name");
+        char *vg_name = get_json_string(lv, "vg_name");
+        char *data_pct_str = get_json_string(lv, "data_percent");
+        char *meta_pct_str = get_json_string(lv, "metadata_percent");
+        if (!(vg_name && lv_name && data_pct_str && meta_pct_str))
+          continue; // get_json_string() would have emitted an error message
+        double data_pct = atof(data_pct_str);
+        double meta_pct = atof(meta_pct_str);
+        lvm_thinp_submit(vg_name, lv_name, data_pct, meta_pct);
+        continue; // skip reporting space allocated in the VG to the thin pool
+                  // as it is counted in the underlying meta/data LVs
+      }
+
+      // Submit the size of the LV as used in the VG
+      char *lv_name = get_json_string(lv, "lv_name");
+      char *vg_name = get_json_string(lv, "vg_name");
+      char *lv_size_str = get_json_string(lv, "lv_size");
+      if (!(vg_name && lv_name && lv_size_str))
+        continue; // get_json_string() would have emitted an error message
+      if (lv_name[0] == '[' && lv_name[strlen(lv_name) - 1] == ']') {
+        // remove [] brackets around names of hidden LVs (eg. thin meta/data)
+        lv_name[strlen(lv_name) - 1] = '\0';
+        lv_name++;
+      }
+      long long int lv_size = atoll(lv_size_str);
+      lvm_vg_submit(vg_name, lv_name, lv_size);
+
+      // Additionally, if it's a snapshot, submit space usage within it
+      if ('s' == lv_attr[0] || 'S' == lv_attr[0]) {
+        char *lv_datap_str = get_json_string(lv, "data_percent");
+        if (lv_datap_str) {
+          double lv_datap = atof(lv_datap_str);
+          lvm_snap_submit(vg_name, lv_name, lv_datap);
+        }
+      } // snapshot
+    }   // foreach LV }}}
+  }     // foreach report
+
+  return 0;
+} // lvm_process_report() }}}
+
+static int lvm_get_report_json(yajl_val *json) // {{{
+{
+  char jsondata[DEFMAXRPTSIZE];
+
+  *json = NULL; // default to error condition (gets set if we make it through)
+
+  // Get the report from lvm into a buffer {{{
+  FILE *fp;
+  char *jsoncmd =
+      "lvm fullreport --all --units=b --nosuffix --reportformat json"
+      " --configreport vg -o vg_name,vg_free"
+      " --configreport pv -S pv_uuid="
+      " --configreport lv -o vg_name,lv_name,lv_size,lv_attr"
+      ",data_percent,metadata_percent"
+      " --configreport pvseg -S pv_uuid="
+      " --configreport seg -S lv_uuid="
+      " 2>/dev/null";
+  size_t rd;
+  jsondata[0] = '\0';
+  if (NULL == (fp = popen(jsoncmd, "r"))) {
+    P_ERROR("popen(): %s", STRERRNO);
```

Comment:
Please don't use *popen* – collectd is a multi-threaded process and spawning children is very tricky. If we can't link with a library and avoid the fork, please re-write as code that can be executed by the [exec plugin](https://collectd.org/wiki/index.php/Plugin:Exec).

Reply:
Ouch. The only supported library that I know which handles lvm is [libblockdev](/storaged-project/libblockdev). It requires GLib (which is why I didn't use it to begin with) and uses GLib's `g_spawn_async_with_pipes()` to do essentially the same thing using either `posix_spawn()` or `fork()`/`exec()`. I presume that would also be unacceptable?

Reply:
The other problem is that the call to `lvm` to get the info requires root access, which I think is incompatible with the exec plugin.

Reply:
@octo  - any advice on how to proceed given the above?

Reply:
Any progress on this topic?